common/tqid/task_queue_id.go

```diff
@@ -0,0 +1,339 @@
+	// a TaskQueue can be broken down to multiple sticky or normal partitions.
+	TaskQueue struct {
+		family   TaskQueueFamily
+		taskType enumspb.TaskQueueType
+	}
+
+	// Partition is a sticky or normal partition of a TaskQueue.
+	// Each Partition has a distinct task queue partition manager in memory in Matching service.
+	// Normal partition with `partitionId=0` is called the "root". Sticky queues are not considered root.
+	Partition interface {
+		NamespaceID() namespace.ID
+		TaskQueue() *TaskQueue
+		TaskType() enumspb.TaskQueueType
+		// IsRoot always returns false for Sticky partitions
+		IsRoot() bool
+		IsSticky() bool
+		Kind() enumspb.TaskQueueKind
+
+		// RpcName returns the mangled name of the task queue partition, to be used in RPCs.
+		//
+		// RPC names look like this:
+		//
+		//  sticky partition:			<sticky name>
+		//	root normal partition: 		<task queue name>
+		//	non-root normal partition: 	/_sys/<task queue name>/<partition id>
+		//
+		// This scheme lets users use anything they like for a base name, except for strings
+		// starting with "/_sys/", without ambiguity.
+		//
+		// For backward compatibility, unversioned low-level task queues with partition 0 do not
+		// use mangled names, they use the bare base name.
+		RpcName() string
+		Key() PartitionKey
+	}
+
+	// NormalPartition is used to distribute load of a TaskQueue in multiple Matching instances. A normal partition is
+	// identified by `partitionId`. The partition with ID 0 is called a root partition.
+	NormalPartition struct {
+		taskQueue   *TaskQueue
+		partitionId int
+	}
+
+	// StickyPartition is made by SDK for a single workflow worker to keep workflow tasks of the same execution
+	// in the same worker for caching benefits. Each sticky partition is identified by a unique `stickyName`
+	// generated by SDK. A StickyPartition can only have workflow task type.
+	StickyPartition struct {
+		stickyName string
+		taskQueue  *TaskQueue
+	}
+
+	// PartitionKey uniquely identifies a task queue partition, to be used in maps.
+	// Note that task queue kind (sticky vs normal) and normal name for sticky task queues are not
+	// part of the task queue partition identity.
+	PartitionKey struct {
+		namespaceID string
+		name        string
+		partitionId int
+		taskType    enumspb.TaskQueueType
+	}
+)
+
+var _ Partition = (*NormalPartition)(nil)
+var _ Partition = (*StickyPartition)(nil)
+
+var (
+	ErrNoParent      = errors.New("root task queue partition has no parent")
+	ErrInvalidDegree = errors.New("invalid task queue partition branching degree")
+	ErrNonZeroSticky = errors.New("only sticky partitions can not have non-zero partition ID")
+)
+
+// FromFamilyName takes a user-provided task queue name (aka family name) and returns a TaskQueueFamily. Returns an
+// error if name looks like a mangled name.
+func FromFamilyName(namespaceId string, name string) (*TaskQueueFamily, error) {
```

Comment:
We should incorporate `TaskQueueFamily` into function name.  Unless namespace will be named `TaskQueueFamily`. When I read code `tqid.FromFamilyName` there is no way I can understand that we create a new `TaskQueueFamily`.